codalab/lib/bundle_store.py

```diff
@@ -80,6 +60,336 @@ def reset(self):
+    """
+
+    # Location where BalancedMultiDiskBundleStore data and temp data is kept relative to CODALAB_HOME
+    DATA_SUBDIRECTORY = 'bundles'
+    TEMP_SUBDIRECTORY = 'temp'
+    CACHE_SIZE = 1 * 1000 * 1000 # number of entries to cache
+    MISC_TEMP_SUBDIRECTORY = 'misc_temp' # BundleServer writes out to here, so should have a different name
+
+    def require_partitions(f):
+        """Decorator added to BalancedMultiDiskBundleStore methods that require a disk to
+        be added to the deployment for tasks to succeed. Prints a helpful error
+        message prompting the user to add a new disk.
+        """
+        def wrapper(*args, **kwargs):
+            self = args[0]
+            if len(self.nodes) < 1:
+                print >> sys.stderr,"""
+    Error: No partitions available.
+    To use BalancedMultiDiskBundleStore, you must add at least one partition. Try the following:
+
+        $ cl help bs-add-partition
+    """
+                sys.exit(1)
+            else:
+                return f(*args, **kwargs)
+        return wrapper
+
+    def __init__(self, codalab_home):
+        self.codalab_home = path_util.normalize(codalab_home)
+
+        self.partitions = os.path.join(self.codalab_home, 'partitions')
+        self.mtemp = os.path.join(self.codalab_home, BalancedMultiDiskBundleStore.MISC_TEMP_SUBDIRECTORY)
+
+        # Perform initialization first to ensure that directories will be populated
+        super(BalancedMultiDiskBundleStore, self).__init__()
+        nodes, _ = path_util.ls(self.partitions)
+        self.nodes = nodes
+        self.lru_cache = OrderedDict()
+        super(BalancedMultiDiskBundleStore, self).__init__()
+
+    def get_node_avail_frac(self, node):
+        # rather than get absolute free space, get fraction
+        st = os.statvfs(node)
+        free = st.f_bavail * st.f_frsize
+        used = (st.f_blocks - st.f_bfree) * st.f_frsize
+        total = st.f_blocks * st.f_frsize
+        return float(free) / total
+
+    @require_partitions
+    def get_bundle_location(self, uuid):
+        """
+        get_bundle_location: look for bundle in the cache, or if not in cache, go through every partition.
+        If not in any partition, return disk with largest free space
+        """
+        if uuid in self.lru_cache:
+            disk = self.lru_cache.pop(uuid)
+        else:
+            disk = None
+            for n in self.nodes: # go through every partition
+                bundle_path = os.path.join(self.partitions, n, BalancedMultiDiskBundleStore.DATA_SUBDIRECTORY, uuid)
+                if os.path.exists(bundle_path):
+                    disk = n
+
+            if disk is None:
+                # return disk with largest free space
+                disk = max(self.nodes, key=lambda x:
+                        self.get_node_avail_frac(os.path.join(self.partitions, x, MultiDiskBundleStore.DATA_SUBDIRECTORY))
```

Comment:
Why wouldn't you return the disk with the max absolute space?  Seems like that's more sensible...